#2 Login 구현

hamryt opened the pull request:
- Interfaces : UserController 
- domain : User, UserRepository 
- application : UserService, EmailExistedException(로그인시 중복 이메일 시 예외 처리)
- config : SecurityJavaConfig ( PasswordEncoder 빈 등록, mysql 접근 가능 처리, 프론트 구현시 필요한 cors 제한 해제)
- 및 JUnit 테스트 코드 포함

hamryt committed:
#2 Login 구현

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/application/UserService.java:
```diff
@@ -0,0 +1,42 @@
+@Service
+public class UserService {
+
+    @Autowired
```
생성자주입을 하므로 이거는 필요없는 라인 같습니다~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/application/UserService.java:
```diff
@@ -0,0 +1,42 @@
+
+    private PasswordEncoder passwordEncoder;
+
+    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
```
이거는 lombok으로 간소화시킬 수 있을 것 같네요~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/domain/User.java:
```diff
@@ -0,0 +1,24 @@
+
+import lombok.*;
+
+import javax.persistence.*;
```
1. 혹시 checkstyle 적용하신걸까요?
2. 만약 안되어있다면 무엇인지 알아보신다음 적용해주세요~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/domain/User.java:
```diff
@@ -0,0 +1,24 @@
+import javax.persistence.*;
+
+
+@Entity
```
mybatis를 쓰는데 jpa 관련 코드가 들어있네요~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/interfaces/UserController.java:
```diff
@@ -0,0 +1,36 @@
+        User user = userService.registerUser(email, name, password);
+
+        String url = "/users/" + user.getId();
+        return ResponseEntity.created(new URI(url)).body("{}");
```
body에 들어가는 값은 의미없는 값 같습니다~ 꼭 넣어줘야할까요?

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/interfaces/UserController.java:
```diff
@@ -0,0 +1,36 @@
+
+    // 회원가입
+    @PostMapping("/users")
+    public ResponseEntity<?> create(
```
제네릭 타입은 명확하게 정해주시는게 좋습니다~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/interfaces/UserController.java:
```diff
@@ -0,0 +1,36 @@
+
+        User user = userService.registerUser(email, name, password);
+
+        String url = "/users/" + user.getId();
```
이런 공통된 템플릿은 따로 분리하는게 재사용성에 좋을 것 같네요~

hamryt committed:
Create join service

to make join service I made 6 packages for categorise class.
1. config
(DatabaseConfiguration: setting mybatis configuration,
SecurityJavaConfig: for connecting with mysql )
2.contoller
(UserController: RestApi for Post /users)
3.dto
(UserDto: setting join  user information)
4. exception
(EmailExistedException: when other join user has same email then throw
this class exception
UnexpectedInsertException: when inserting not acted then throw this
class exception)
5. mapper
(UserMapper: mapper for userDto and UserMapper.xml)
6. service
(UserServiceImpl: has business logic about user
UserService: interface for UserServiceImpl)!

hamryt committed:
reate join service

to make join service I made 6 packages for categorise class.
1. config
(DatabaseConfiguration: setting mybatis configuration,
SecurityJavaConfig: for connecting with mysql )
2.contoller
2.contoller
(UserController: RestApi for Post /users)
3.dto
3.dto
(UserDto: setting join  user information)
4. exception
(EmailExistedException: when other join user has same email then throw
this class exception
UnexpectedInsertException: when inserting not acted then throw this
class exception)
5. mapper
(UserMapper: mapper for userDto and UserMapper.xml)
6. service
(UserServiceImpl: has business logic about user
UserService: interface for UserServiceImpl)

hamryt committed:
Create login service

hamryt made login service.
1. config
(DatabaseConfiguration: setting mybatis configuration )
2.contoller
(UserController: RestApi for Post /users)
3. dto
(UserDto: user information)
4. exception
(EmailExistedException: when other join user has same email then throw
this class exception
UnexpectedInsertException: when inserting not activated then throw this
class exception)
5.mapper
(UserMapper: mapper for UserMapper.xml)
6.service
(UserServiceImpl: has business logic about user
UserService: interface for UserServiceImpl)
7.util
(SecurityUtil: for sha256)

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final UserService userService;
+
+    // 회원가입
```
이미 메소드명으로 회원가입이라는건 유추가 가능합니다~ 필요없는 주석 같네요

hamryt replied:
넵 삭제하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@AllArgsConstructor
```
`@RequiredArgConstructor`와 이것의 차이는 무엇일까요?

hamryt replied:
@AllArgsConstructor은 클래스의 모든 변수를 인자로 사용하는 생성자를 만들고 
@RequiredArgContsructor는 특정 필요한 변수만을 인자로 하는 생성자를 만듭니다. 
제가 작성한 클래스는 final을 사용한 변수를 사용했기에 @RequiredArgConstructor를 사용하는 것이 알맞을 것 같습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/mapper/UserMapper.java:
```diff
@@ -0,0 +1,14 @@
+
+    int insertUser(UserDto user);
+
+    Boolean isExistsEmail(String email);
```
여기서는 null이 리턴될 수 있나요? 그렇지 않다면 primitive type을 쓰는게 좋아보입니다~

hamryt replied:
넵 null은 리턴 되지 않으므로 수정하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+            userMapper.insertUser(newUser);
+        } catch (RuntimeException e) {
+            log.error(e.toString());
+            throw new UnexpectedInsertException(userDto.toString());
```
이거 같은경우는 예외가 발생하지도 않았는데 insert row가 1개가 아닐때 발생하는 상황이 아닐까요?

hamryt replied:
넵  의미상 if (userMapper.insertUser(newUser) == 0) 인 것이 더 명확하기에 다시 바꾸겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+        try {
+            userMapper.insertUser(newUser);
+        } catch (RuntimeException e) {
+            log.error(e.toString());
```
toString()보다는 예외 객체를 통째로 넘기는게 좋습니다~ `logger.error` 메소드의 시그니처가 어떻게 되는지 알아보시면 좋을 것 같네요~

hamryt replied:
찾아보니 Logger.error(e.getMessage(), e); 이렇게 사용한 다는 걸 알게되었습니다. 지금 제 코드에선 뺐지만 다음엔 이렇게 사용하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/utill/SecurityUtil.java:
```diff
@@ -0,0 +1,7 @@
+package com.hamryt.helparty.utill;
+
+public interface SecurityUtil {
```
자바에서 유틸클래스가 어떤 것을 의미하는지 알아보시면 좋을 것 같습니다~

hamryt replied:
자바에서 유틸클래스는 반복적이고 공통으로 쓰이는 기능을 처리할 때 사용하는 클래스입니다. 
암호화하는 클래스는 공통으로 반복적으로 쓰이는 클래스이므로 이름을 어느 한 방법에 종속적이지 않게 포괄적인 이름으로 짓겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/utill/SecurityUtil.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface SecurityUtil {
+
+    String encryptSha256(String data);
```
암호화 방식에 대한 추상화같은데 이렇게 하면 `sha256`에 종속적이게 되지 않을까요?
만약 다른 방식의 암호화가 추가된다면 이거를 오버라이딩해서 구현하지 못하게 될테니까요~

hamryt replied:
넵 그렇게 생각합니다. 확장에 용이하도록 이름을 범용적으로 바꾸겠습니다.

f-lab-dev commented:
이 브랜치의 리뷰를 반영한 것은 이 브랜치에 push해주세요~
